### PR TITLE
Step1

### PR DESCRIPTION
**Pull request by tomerorian**
Adds the main activity. The activity is responsible for fetching and displaying all the currently playing movies.

image

### CODE REVIEW

**Comment by ifeins** [comment]
One habit we keep is to put screenshots/animated gifs in the description of the PR.
This gives context to what the PR is about and makes it easier to understand the code.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.ArrayList;
+
+
+public class MovieAdapter extends RecyclerView.Adapter<MovieViewHolder> {
+
+    private Context mContext;
+    private ArrayList<NowPlayingResponse.MovieInfo> mMovies;
+
+    public MovieAdapter(Context context) {
+        this.mContext = context;
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: no need for `this` if you're using the `m` notation

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.moviefiend.torian.moviefiend.network.NowPlayingResponse;
+
+import java.util.ArrayList;
+
+
+public class MovieAdapter extends RecyclerView.Adapter<MovieViewHolder> {
+
+    private Context mContext;
+    private ArrayList<NowPlayingResponse.MovieInfo> mMovies;
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: You should always use the interface/base class for the reference type (e.g. `List` and not `ArrayList`).

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    public MovieAdapter(Context context) {
+        this.mContext = context;
+
+        mMovies = new ArrayList<>();
+    }
+
+    public void setMovies(ArrayList<NowPlayingResponse.MovieInfo> movies) {
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: `List`
Here it's especially important since you don't to force the users of this method to pass a concrete implementation, they should be able to pass a `LinkedList` if they wish.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieViewHolder.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package com.moviefiend.torian.moviefiend;
+
+import android.support.v7.widget.RecyclerView;
+import android.view.View;
+import android.widget.ImageView;
+import android.widget.TextView;
+
+import com.squareup.picasso.Picasso;
+
+public class MovieViewHolder extends RecyclerView.ViewHolder {
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieViewHolder.java:11]: Any special reason to define it as a top level class and not an inner class of the adapter?
It's only used in the adapter after all

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        return new MovieViewHolder(itemView);
+    }
+
+    @Override
+    public void onBindViewHolder(MovieViewHolder holder, int position) {
+        NowPlayingResponse.MovieInfo movie = mMovies.get(position);
+        holder.setTitle(movie.getTitle());
+        holder.setPosterView(mContext.getString(R.string.tmdb_poster_url, movie.getPosterPath()));
+        holder.setRating(movie.getRating());
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: Instead of this you can just have the view holder implement a `setMovie` method and it internally will assign all the relevant child views.

This way if in the future you'll want to reuse the view holder class then you won't need to duplicate this code and simply pass a `Movie` object.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieInfoLoader.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    protected void onForceLoad() {
+        loadMovies();
+    }
+
+    @Override
+    protected void onStartLoading() {
+        loadMovies();
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieInfoLoader.java:28]: `onForceLoad`

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    public static final int MOVIE_LOADER_ID = 0;
+
+    private RecyclerView mRecyclerView;
+    private MovieAdapter mMovieAdapter;
+
+    @Override
+    public void onActivityCreated(Bundle savedInstanceState) {
+        super.onActivityCreated(savedInstanceState);
+
+        getLoaderManager().initLoader(MOVIE_LOADER_ID, null, this);
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java:28]: I'm not 100% sure but I think that if you call `initLoader` from `onActivityCreated` the `onLoadFinished` method will be called twice.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Nullable
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java:31]: Curious, why did you add the `Nullable` annotation specifically here?

File: app/src/main/java/com/moviefiend/torian/moviefiend/network/NowPlayingResponse.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package com.moviefiend.torian.moviefiend.network;
+
+import com.google.gson.annotations.SerializedName;
+
+import java.util.ArrayList;
+
+public class NowPlayingResponse {
+    @SerializedName("results")
+    private ArrayList<MovieInfo> movies;
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/network/NowPlayingResponse.java]: If you're already using `@SerializedName` then you might as well use the `m` notation for the member variables.

File: app/src/main/java/com/moviefiend/torian/moviefiend/network/NowPlayingResponse.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    public ArrayList<MovieInfo> getMovies() {
+        return movies;
+    }
+
+    public static class MovieInfo {
+        @SerializedName("poster_path")
+        String posterPath;
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/network/NowPlayingResponse.java]: same here

File: app/src/main/res/layout/movie_item_view.xml [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:orientation="vertical"
+        android:layout_toRightOf="@+id/poster">
+
+        <TextView
+            android:id="@+id/title"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            tools:text="Kings Lending" />
```
Comment (ifeins) [inline app/src/main/res/layout/movie_item_view.xml:26]: 👍

File: app/src/main/res/layout/movie_list_fragment.xml [hunk 12]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (ifeins) [inline app/src/main/res/layout/movie_list_fragment.xml:2]: If it only contains one view, any reason to use `LinearLayout` and not `FrameLayout`?
`FrameLayout` is cheaper in terms of initialization and performance.

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -1,3 +1,8 @@
 <resources>
     <string name="app_name">Movie Fiend</string>
+    <string name="poster">poster</string>
+    <string name="more_details">more details</string>
+    <string name="tmdb_now_playing_url">http://api.themoviedb.org/3/movie/now_playing?api_key=%s</string>
```
Comment (ifeins) [inline app/src/main/res/values/strings.xml]: Add `translatable="false"` to indicate that this string should not be localized.
Especially important in Venice where there are localization teams that translate the strings added to the project.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+import android.support.v7.widget.RecyclerView;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import com.moviefiend.torian.moviefiend.network.NowPlayingResponse;
+
+import java.util.ArrayList;
+
+public class MovieListFragment extends Fragment implements LoaderManager.LoaderCallbacks<ArrayList<NowPlayingResponse.MovieInfo>> {
```
Comment (ifeins) [inline app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java:18]: Any special reason why you decide to make it a fragment (besides reusing it in step 6 :))?

That's fine for the training, but one thing to consider is that now you have an activity and its layout that are completely empty.
So in my opinion you're not gaining much by having both an activity and a fragment.

**Review by ifeins — COMMENTED** [review]
Very clean, good job 👍 
Some small comments.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java (hunk 2 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: you are right kind sir

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java (hunk 1 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: ack, auto-generated

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java (hunk 5 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: @ifeins imo its kind of "mazgan" (preparing ahead for something I don't need). no one guarantees that the next time I might use it I will have a `MovieInfo` object.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieViewHolder.java (hunk 4 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieViewHolder.java:11]: @ifeins code cleaning-ness. do you usually define those in the same class?

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieInfoLoader.java (hunk 6 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieInfoLoader.java:28]: @ifeins why call onForceLoad and not the actual method I want to call?

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java (hunk 14 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java:18]: @ifeins just what I'm used to. as we talked today.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java (hunk 7 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java:28]: @ifeins I just debugged and its not. what other options do I have?

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java (hunk 8 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java:31]: @ifeins I didn't, its automatically added by Android Studio because the super method declares it

File: app/src/main/java/com/moviefiend/torian/moviefiend/network/NowPlayingResponse.java (hunk 9 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/network/NowPlayingResponse.java]: ack

File: app/src/main/java/com/moviefiend/torian/moviefiend/network/NowPlayingResponse.java (hunk 10 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/network/NowPlayingResponse.java]: ack

File: app/src/main/res/layout/movie_list_fragment.xml (hunk 12 above)
Comment (tomerorian) [reply app/src/main/res/layout/movie_list_fragment.xml:2]: @ifeins that was the default layout created, can you point me to resources about the performance thing? I searched (quickly) and didn't find anything

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (tomerorian) [reply app/src/main/res/values/strings.xml]: ack, didn't know about it. thanks

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java (hunk 7 above)
Comment (ifeins) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieListFragment.java:28]: @tomerorian Can you check how many times `onLoadFinished` gets called when you rotate the device when waiting for the network request?
It should be called only once.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java (hunk 5 above)
Comment (ifeins) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: I agree that the potential for reuse in other places is small, but we're not just doing it for re-use.
It also cleans up the code of the view holder that you just pass into it the domain model `Movie` and it know internally how to display it.
This article sums it up really nicely: https://jonfhancock.com/your-viewholders-are-dumb-make-em-not-dumb-82e6f73f630c#.3mj413nri

In addition if you add new methods to `Movie` you will need to add new methods to the view holder.

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieViewHolder.java (hunk 4 above)
Comment (ifeins) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieViewHolder.java:11]: Yes, usually we place them as nested class.

File: app/src/main/res/layout/movie_list_fragment.xml (hunk 12 above)
Comment (ifeins) [reply app/src/main/res/layout/movie_list_fragment.xml:2]: hmm, I actually can't find any reference to it.
Feels to me that you should use the simplest layout that satisfy your design requirements (i.e. not use a `RelativeLayout` if you all you need is to center a single view) but it's just a hunch, can't really back it up with proof :)

**Comment by ifeins** [comment]
A few other remaining comments.
After fixing them then merge away :)

File: app/src/main/res/layout/movie_list_fragment.xml (hunk 12 above)
Comment (tomerorian) [reply app/src/main/res/layout/movie_list_fragment.xml:2]: @ifeins from what I saw `FrameLayout` is there to stack views on top of each other, so it doesn't sound better fitting than `LinearLayout`

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java (hunk 5 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieAdapter.java]: @ifeins I'm not yet fully convinced about this method, but for the sake of this training I will use it :)

File: app/src/main/java/com/moviefiend/torian/moviefiend/MovieViewHolder.java (hunk 4 above)
Comment (tomerorian) [reply app/src/main/java/com/moviefiend/torian/moviefiend/MovieViewHolder.java:11]: @ifeins critical? especially now that the view holder has some "logic" I rather keep it clean
